File: python/15-inheritance-abuse/src/employee.py
class Employee:
    def __init__(self, name, salary): self.name=name; self.salary=salary
    def calculate_bonus(self): return self.salary * 0.02
    def calculate_total_reward(self, performance, years):
        x = self.calculate_bonus()
        if performance == "high": x += self.salary * 0.01
        if years >= 5: x += 500
        return x
class Manager(Employee):
    def calculate_bonus(self): return self.salary * 0.05
    def calculate_total_reward(self, performance, years):
        x = self.calculate_bonus()
        if performance == "high": x += self.salary * 0.02
        if years >= 5: x += 1000
        return x
class SeniorManager(Manager):
    def calculate_bonus(self):
        base = self.salary * 0.05
        if base > 10000: base = 10000
        return base
    def calculate_total_reward(self, performance, years):
        x = self.calculate_bonus()
        if performance == "high": x += self.salary * 0.02
        if years >= 5: x += 1500
        return x
class Director(SeniorManager):
    def calculate_bonus(self):
        base = self.salary * 0.05
        if base > 20000: base = 20000
        return base
    def calculate_total_reward(self, performance, years):
        x = self.calculate_bonus()
        if performance == "high": x += self.salary * 0.03
        if years >= 5: x += 2500
        return x

File: ruby/15-inheritance-abuse/src/employee.rb
# frozen_string_literal: true

class Employee
  def initialize(name, salary)
    @name = name
    @salary = salary
  end

  def calculate_bonus
    @salary * 0.02
  end

  def calculate_total_reward(performance, years)
    x = calculate_bonus
    x += @salary * 0.01 if performance == 'high'
    x += 500 if years >= 5
    x
  end
end

class Manager < Employee
  def calculate_bonus
    @salary * 0.05
  end

  def calculate_total_reward(performance, years)
    x = calculate_bonus
    x += @salary * 0.02 if performance == 'high'
    x += 1000 if years >= 5
    x
  end
end

class SeniorManager < Manager
  def calculate_bonus
    base = @salary * 0.05
    base > 10_000 ? 10_000 : base
  end

  def calculate_total_reward(performance, years)
    x = calculate_bonus
    x += @salary * 0.02 if performance == 'high'
    x += 1500 if years >= 5
    x
  end
end

class Director < SeniorManager
  def calculate_bonus
    base = @salary * 0.05
    base > 20_000 ? 20_000 : base
  end

  def calculate_total_reward(performance, years)
    x = calculate_bonus
    x += @salary * 0.03 if performance == 'high'
    x += 2500 if years >= 5
    x
  end
end
